src/modules/ekf2/EKF/output_predictor.cpp

```diff
@@ -48,126 +48,183 @@ void OutputPredictor::print_status()
+	_R_to_earth_now = Dcmf(_output_new.quat_nominal);
+
+	// apply change to velocity
+	_output_new.vel = quat_change.rotateVector(_output_new.vel);
+	_output_new.vel_alternative = quat_change.rotateVector(_output_new.vel_alternative);
+
+	propagateVelocityUpdateToPosition();
 }
 
-void OutputPredictor::resetHorizontalVelocityTo(const Vector2f &delta_horz_vel)
+void OutputPredictor::resetHorizontalVelocityTo(const Vector2f &new_horz_vel)
 {
-	for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
-		_output_buffer[index].vel.xy() += delta_horz_vel;
+	const outputSample &output_delayed = _output_buffer.get_oldest();
+
+	// horizontal velocity
+	{
+		const Vector2f delta_vxy = new_horz_vel - Vector2f(output_delayed.vel.xy());
+
+		for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
+			_output_buffer[index].vel.xy() += delta_vxy;
+		}
+
+		_output_new.vel.xy() += delta_vxy;
 	}
 
-	_output_new.vel.xy() += delta_horz_vel;
+	// horizontal velocity (alternative)
+	{
+		const Vector2f delta_vxy = new_horz_vel - Vector2f(output_delayed.vel_alternative.xy());
+
+		for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
+			_output_buffer[index].vel_alternative.xy() += delta_vxy;
+		}
+
+		_output_new.vel_alternative.xy() += delta_vxy;
+	}
+
+	// propagate position forward using the reset velocity
+	propagateVelocityUpdateToPosition();
 }
 
-void OutputPredictor::resetVerticalVelocityTo(float delta_vert_vel)
+void OutputPredictor::resetVerticalVelocityTo(const float new_vert_vel)
 {
-	for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
-		_output_buffer[index].vel(2) += delta_vert_vel;
-		_output_vert_buffer[index].vert_vel += delta_vert_vel;
+	const outputSample &output_delayed = _output_buffer.get_oldest();
+
+	// vertical velocity
+	{
+		const float delta_z = new_vert_vel - output_delayed.vel(2);
+
+		for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
+			_output_buffer[index].vel(2) += delta_z;
+		}
+
+		_output_new.vel(2) += delta_z;
+	}
+
+	// vertical velocity (alternative)
+	{
+		const float delta_z = new_vert_vel - output_delayed.vel_alternative(2);
+
+		for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
+			_output_buffer[index].vel_alternative(2) += delta_z;
+		}
+
+		_output_new.vel_alternative(2) += delta_z;
 	}
 
-	_output_new.vel(2) += delta_vert_vel;
-	_output_vert_new.vert_vel += delta_vert_vel;
+	// propagate vertical position forward using the reset velocity
+	propagateVelocityUpdateToPosition();
 }
 
-void OutputPredictor::resetHorizontalPositionTo(const Vector2f &delta_horz_pos)
+void OutputPredictor::resetHorizontalPositionTo(const Vector2f &new_horz_pos)
 {
-	for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
-		_output_buffer[index].pos.xy() += delta_horz_pos;
+	const outputSample &output_delayed = _output_buffer.get_oldest();
+
+	// horizontal velocity
```

Comment:
```suggestion
	// horizontal position
```

or remove it, I'm not convinced that it really helps, even if it is correct